### ISSUE TITLE
[WebKit] App crashes when decidePolicy is implemented on generic class with dynamic registrar (simulator)

### ISSUE THREAD

**Issue by wjosset** [xamarin/xamarin-macios#6687]
### Steps to Reproduce

1. Sub-class a ViewController and make it generic
2. Make it implement the IWKNavigationDelegate (using WebKit)
3. Implement the decidePolicy delegate method within this generic class
4. Run app on simulator (which is by default using dynamic registrar)

### Expected Behavior
The application should run

### Actual Behavior
The application is crashing at runtime

### Remarks
The problem is not occuring when:
1. Using other delegate methods such as "didCommitNavigation"
2. Generic part of the class is taken out
3. Device is targeted (which has a static registrar by default)
4. Forcing simulator to use static registrar (iOS Build => additional mtouch arguments => --registrar:static)

The problem is solved on simulator but occurs ON DEVICE when:
1. The implementation of "decidePolicy" method goes to the child class. This problem occurs even if the phone iOS build registrar is setted to dynamic (like the simulator). Here is the raised exeption =>

```
ObjCRuntime.RuntimeException
Unable to locate the block to delegate conversion method for the method
DecidePolicy's parameter #3
```

### Environment

```
=== Visual Studio Community 2019 for Mac ===

Version 8.2 (build 1105)
Installation UUID: fde44cf8-6019-4f41-97ea-fc050c633eed
	GTK+ 2.24.23 (Raleigh theme)
	Xamarin.Mac 5.14.0.85 (d16-2-fix-mac-updates / 86973979)

	Package version: 600000311

=== Mono Framework MDK ===

Runtime:
	Mono 6.0.0.311 (2019-02/494641b300c) (64-bit)
	Package version: 600000311

=== NuGet ===

Version: 5.1.0.6013

=== .NET Core ===

Runtime: /usr/local/share/dotnet/dotnet
Runtime Versions:
	2.1.12
	2.1.11
	2.1.9
SDK: /usr/local/share/dotnet/sdk/2.1.701/Sdks
SDK Versions:
	2.1.701
	2.1.700
	2.1.505
MSBuild SDKs: /Library/Frameworks/Mono.framework/Versions/6.0.0/lib/mono/msbuild/Current/bin/Sdks

=== Xamarin.Profiler ===

Version: 1.6.11
Location: /Applications/Xamarin Profiler.app/Contents/MacOS/Xamarin Profiler

=== Updater ===

Version: 11

=== Apple Developer Tools ===

Xcode 10.2.1 (14490.122)
Build 10E1001

=== Xamarin.Mac ===

Version: 5.14.0.110 (Visual Studio Community)
Hash: a8bcecce
Branch: d16-2
Build date: 2019-06-28 07:31:43-0400

=== Xamarin.iOS ===

Version: 12.14.0.110 (Visual Studio Community)
Hash: a8bcecce
Branch: d16-2
Build date: 2019-06-28 07:31:42-0400

=== Xamarin Designer ===

Version: 16.2.0.354
Hash: 5f80a14f8
Branch: remotes/origin/d16-2
Build date: 2019-07-11 17:28:26 UTC

=== Xamarin.Android ===

Version: 9.4.0.51 (Visual Studio Community)
Commit: xamarin-android/d16-2/9fa7775
Android SDK: /Users/wj/Library/Developer/Xamarin/android-sdk-macosx
	Supported Android versions:
		6.0 (API level 23)
		7.1 (API level 25)
		8.1 (API level 27)

SDK Tools Version: 26.1.1
SDK Platform Tools Version: 28.0.0
SDK Build Tools Version: 27.0.3

Build Information: 
Mono: mono/mono/2019-02@e6f5369c2d2
Java.Interop: xamarin/java.interop/d16-2@d64ada5
LibZipSharp: grendello/LibZipSharp/d16-2@caa0c74
LibZip: nih-at/libzip/rel-1-5-1@b95cf3f
ProGuard: xamarin/proguard/master@905836d
SQLite: xamarin/sqlite/3.27.1@8212a2d
Xamarin.Android Tools: xamarin/xamarin-android-tools/d16-2@6f6c969

=== Microsoft Mobile OpenJDK ===

Java SDK: /Users/wj/Library/Developer/Xamarin/jdk/microsoft_dist_openjdk_1.8.0.25
1.8.0-25
Android Designer EPL code available here:
https://github.com/xamarin/AndroidDesigner.EPL

=== Android Device Manager ===

Version: 1.2.0.58
Hash: d03e3ea
Branch: remotes/origin/d16-2~1
Build date: 2019-07-16 17:37:14 UTC

=== Xamarin Inspector ===

Version: 1.4.3
Hash: db27525
Branch: 1.4-release
Build date: Mon, 09 Jul 2018 21:20:18 GMT
Client compatibility: 1

=== Build Information ===

Release ID: 802001105
Git revision: a187f6852ae62ba52b944b47e53702957edfcf7b
Build date: 2019-07-22 18:20:10+00
Build branch: release-8.2
Xamarin extensions: 34bd2378d8d707df25989727c23ce68d9a232a67

=== Operating System ===

Mac OS X 10.14.5
Darwin 18.6.0 Darwin Kernel Version 18.6.0
    Thu Apr 25 23:16:27 PDT 2019
    root:xnu-4903.261.4~2/RELEASE_X86_64 x86_64


```

### Build Logs

```
2019-07-30 14:03:56.883716+0200 testregistrar[4540:184363] ObjCRuntime.RuntimeException: The registrar found an invalid generic parameter type 'System.Action`1[[WebKit.WKNavigationActionPolicy, Xamarin.iOS, Version=0.0.0.0, Culture=neutral, PublicKeyToken=84e04ff9cfb79065]]' in the parameter decisionHandler of the method 'testregistrar.GenericWebViewController`1.DecidePolicy(WebKit.WKWebView,WebKit.WKNavigationAction,System.Action`1[[WebKit.WKNavigationActionPolicy, Xamarin.iOS, Version=0.0.0.0, Culture=neutral, PublicKeyToken=84e04ff9cfb79065]])'. The generic parameter must have an 'NSObject' constraint.
```

### Example Project

https://gist.github.com/wjosset/a8b1a7e8590bb40ac0071f804507f350

```
using System;
using Foundation;
using UIKit;
using WebKit;

namespace testregistrar
{
    public class ProjectAbstractViewController : UIViewController
    {
        protected ProjectAbstractViewController(IntPtr handle)
            : base(handle)
        {
        }
    }

    public class GenericWebViewController<WebViewModel> : ProjectAbstractViewController, IWKNavigationDelegate
    {
        private WKWebView _webView;

        protected GenericWebViewController(IntPtr handle)
            : base(handle)
        {
        }

        public override void ViewDidLoad()
        {
            base.ViewDidLoad();
            SetupWebView();
            LoadRequest();
        }

        private void SetupWebView()
        {
            _webView = new WKWebView(View.Bounds, new WKWebViewConfiguration());
            _webView.NavigationDelegate = this;
            View.AddSubview(_webView);
        }

        private void LoadRequest()
        {
            var url = "https://www.google.com";
            var nsUrl = new NSUrl(url);
            var nsUrlRequest = new NSUrlRequest(nsUrl);
            _webView.LoadRequest(nsUrlRequest);
        }

        [Export("webView:decidePolicyForNavigationAction:decisionHandler:")]
        public void DecidePolicy(WKWebView webView, WKNavigationAction navigationAction, System.Action<WKNavigationActionPolicy> decisionHandler)
        {
            Console.WriteLine("decidePolicy Triggered");
            decisionHandler(WKNavigationActionPolicy.Allow);
        }


        [Export("webView:didCommitNavigation:")]
        public void DidCommitNavigation(WKWebView webView, WKNavigation navigation)
        {
            Console.WriteLine("DidCommitNavigation Triggered");
        }
    }

    public class WebViewModel
    {
    }

    public partial class ViewController : GenericWebViewController<WebViewModel>
    {
        protected ViewController(IntPtr handle)
        : base(handle)
        {
        }
    }
}
```

**Comment by chamons** [comment]
FYI - @rolfbjarne

**Comment by EbenDang** [comment]
I go the same issue when running an app in simulator:

`
ObjCRuntime.RuntimeException: The registrar found an invalid generic parameter type 'System.Action`1[[WebKit.WKNavigationActionPolicy, Xamarin.iOS, Version=0.0.0.0, Culture=neutral, PublicKeyToken=84e04ff9cfb79065]]' in the parameter decisionHandler of the method 'WKWebNavigationDelegateImpl.DecidePolicy(WebKit.WKWebView,WebKit.WKNavigationAction,System.Action`1[[WebKit.WKNavigationActionPolicy, Xamarin.iOS, Version=0.0.0.0, Culture=neutral, PublicKeyToken=84e04ff9cfb79065]])'. The generic parameter must have an 'NSObject' constraint.
`

**Reproduce step**

- Create a simple ios app and override the DecidePolicy function in WKNavigationDelegate
- Build  the app successfully.
- Run it in simulator and app crashed at the very beginning. Got the above runtime exception message.

**Dev environment** 
`=== Visual Studio Community 2019 for Mac ===

Version 8.2.2 (build 21)
Installation UUID: 1d2465a1-341c-4903-a92c-0bb89ae8cafc
	GTK+ 2.24.23 (Raleigh theme)
	Xamarin.Mac 5.14.0.85 (d16-2-fix-mac-updates / 86973979)

	Package version: 600000319

=== Mono Framework MDK ===

Runtime:
	Mono 6.0.0.319 (2019-02/52203786470) (64-bit)
	Package version: 600000319

=== NuGet ===

Version: 5.1.0.6013

=== .NET Core ===

Runtime: /usr/local/share/dotnet/dotnet
Runtime Versions:
	2.1.12
	2.1.9
	2.1.8
SDK: /usr/local/share/dotnet/sdk/2.1.701/Sdks
SDK Versions:
	2.1.701
	2.1.505
	2.1.504
MSBuild SDKs: /Library/Frameworks/Mono.framework/Versions/6.0.0/lib/mono/msbuild/Current/bin/Sdks

=== Xamarin.Profiler ===

Version: 1.6.11
Location: /Applications/Xamarin Profiler.app/Contents/MacOS/Xamarin Profiler

=== Updater ===

Version: 11

=== Apple Developer Tools ===

Xcode 10.3 (14492.2)
Build 10G8

=== Xamarin.Mac ===

Xamarin.Mac not installed. Can't find /Library/Frameworks/Xamarin.Mac.framework/Versions/Current/Version.

=== Xamarin.iOS ===

Version: 12.2.1.16 (Visual Studio Community)
Hash: 2dc06c71
Branch: d15-9
Build date: 2019-02-12 23:09:50-0500

=== Xamarin Designer ===

Version: 16.2.0.379
Hash: 3f0652957
Branch: remotes/origin/d16-2
Build date: 2019-08-01 17:32:50 UTC

=== Xamarin.Android ===

Version: 9.4.1.0 (Visual Studio Community)
Commit: xamarin-android/d16-2/ebd65c9
Android SDK: /Users/yanbodang/Library/Developer/Xamarin/android-sdk-macosx
	Supported Android versions:
		8.1 (API level 27)

SDK Tools Version: 26.1.1
SDK Platform Tools Version: 28.0.0
SDK Build Tools Version: 27.0.3

Build Information: 
Mono: mono/mono/2019-02@e6f5369c2d2
Java.Interop: xamarin/java.interop/d16-2@d64ada5
LibZipSharp: grendello/LibZipSharp/d16-2@caa0c74
LibZip: nih-at/libzip/rel-1-5-1@b95cf3f
ProGuard: xamarin/proguard/master@905836d
SQLite: xamarin/sqlite/3.27.1@8212a2d
Xamarin.Android Tools: xamarin/xamarin-android-tools/d16-2@6f6c969

=== Microsoft Mobile OpenJDK ===

Java SDK: /Users/yanbodang/Library/Developer/Xamarin/jdk/microsoft_dist_openjdk_1.8.0.9
1.8.0-9
Android Designer EPL code available here:
https://github.com/xamarin/AndroidDesigner.EPL

=== Android Device Manager ===

Version: 1.2.0.58
Hash: d03e3ea
Branch: remotes/origin/d16-2~1
Build date: 2019-07-30 21:29:36 UTC

=== Xamarin Inspector ===

Version: 1.4.3
Hash: db27525
Branch: 1.4-release
Build date: Mon, 09 Jul 2018 21:20:18 GMT
Client compatibility: 1

=== Build Information ===

Release ID: 802020021
Git revision: d2ffd09000c5d24a206b715b63d409e8e965689a
Build date: 2019-08-02 00:25:49+00
Build branch: release-8.2
Xamarin extensions: 4cc9cbbe68baaeb6bc654f36e057b1bd124f3f65

=== Operating System ===

Mac OS X 10.14.6
Darwin 18.7.0 Darwin Kernel Version 18.7.0
    Thu Jun 20 18:42:21 PDT 2019
    root:xnu-4903.270.47~4/RELEASE_X86_64 x86_64

=== Enabled user installed extensions ===

StyleCop Support 1.1.0.3
`

**Comment by rolfbjarne** [comment]
Fixed here: https://github.com/xamarin/xamarin-macios/commit/73ce0f888295aa8b0f7bdb82c4c92ea6c092ac98
